libeos-payg/fake-clock.c

```diff
@@ -0,0 +1,167 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright © 2018 Endless Mobile, Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * Alternatively, the contents of this file may be used under the terms of the
+ * GNU Lesser General Public License Version 2.1 or later (the "LGPL"), in
+ * which case the provisions of the LGPL are applicable instead of those above.
+ * If you wish to allow use of your version of this file only under the terms
+ * of the LGPL, and not to allow others to use your version of this file under
+ * the terms of the MPL, indicate your decision by deleting the provisions
+ * above and replace them with the notice and other provisions required by the
+ * LGPL. If you do not delete the provisions above, a recipient may use your
+ * version of this file under the terms of either the MPL or the LGPL.
+ */
+
+#include <libeos-payg/fake-clock.h>
+#include <libeos-payg/boottime-source.h>
+
+#define MSEC_PER_SEC 1000
+
+struct _EpgFakeClock
+{
+  GObject parent_instance;
+
+  gint64 time; /* in seconds */
+  gint64 wallclock_time; /* in seconds */
+};
+
+static void clock_iface_init (EpgClockInterface *iface,
+                              gpointer           iface_data);
+
+G_DEFINE_TYPE_WITH_CODE (EpgFakeClock, epg_fake_clock, G_TYPE_OBJECT,
+                         G_IMPLEMENT_INTERFACE (EPG_TYPE_CLOCK, clock_iface_init))
+
+static gint64
+epg_fake_clock_get_wallclock_time (EpgClock *clock)
+{
+  EpgFakeClock *self = (EpgFakeClock *)clock;
+
+  g_return_val_if_fail (EPG_IS_FAKE_CLOCK (self), -1);
+
+  if (self->wallclock_time == 0)
+    self->wallclock_time = g_get_real_time () / G_USEC_PER_SEC;
+
+  return self->wallclock_time;
+}
+
+/**
+ * epg_fake_clock_set_wallclock_time:
+ * @self: an #EpgFakeClock
+ * @new_time: a positive value to set the clock to
+ *
+ * Set the time to @new_time so that value will be returned by
+ * epg_clock_get_wallclock_time() until the next call to this function. This is
+ * intended to be used by unit tests.
+ *
+ * Since: 0.2.1
+ */
+void
+epg_fake_clock_set_wallclock_time (EpgFakeClock *self,
+                                   gint64        new_time)
+{
+  g_return_if_fail (EPG_IS_FAKE_CLOCK (self));
+  g_return_if_fail (new_time > 0);
+
+  self->wallclock_time = new_time;
+}
+
+static gint64
+epg_fake_clock_get_time (EpgClock *clock)
+{
+  EpgFakeClock *self = (EpgFakeClock *)clock;
+
+  g_return_val_if_fail (EPG_IS_FAKE_CLOCK (self), -1);
+
+  if (self->time == 0)
+    self->time = epg_get_boottime () / G_USEC_PER_SEC;
```

Comment:
The idea was that the tests should use realistic values to test realistically. You could still use an explicit value in a test if you want, by either passing it to `epg_fake_clock_new` or to `set_time`/`set_wallclock_time`. Does it matter if the unit tests use slightly different values each time?

Reply:
You can use realistic values and still have the tests be deterministic, for example, by choosing static clock values from 2018 rather than the bottom of the range of valid integers.

> Does it matter if the unit tests use slightly different values each time?

As a general principle, yes, this makes for hard to reproduce failures in future. If the test inputs change and the code under test changes, then the test fails — what’s caused the failure?

Reply:
I implemented this in a fixup